### ISSUE TITLE
Error when setting up West Berks Council

### ISSUE THREAD

**Issue by inspoadam** [robbrad/UKBinCollectionData#1040]
### Home Assistant Version

2024.11.3

### Installation Method

Using HACS

### Name of Council (if relevant)

West Berks Council 

### Issue Information

When setting up the above council, I receive an error as follows. The setup continues in a loop and no entities are created.

Error trace...

```
This error originated from a custom integration.

Logger: custom_components.uk_bin_collection
Source: custom_components/uk_bin_collection/__init__.py:250
integration: UK Bin Collection Data (documentation, issues)
First occurred: 22 November 2024 at 23:43:17 (32 occurrences)
Last logged: 00:05:16

[UKBinCollection] Unexpected error: [Errno 8] Exec format error: '/root/.wdm/drivers/chromedriver/linux64/114.0.5735.90/chromedriver'
Traceback (most recent call last):
  File "/config/custom_components/uk_bin_collection/__init__.py", line 250, in _async_update_data
    data = await asyncio.wait_for(
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/asyncio/tasks.py", line 520, in wait_for
    return await fut
           ^^^^^^^^^
  File "/usr/local/lib/python3.12/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/uk_bin_collection/uk_bin_collection/collect_data.py", line 101, in run
    return self.client_code(
           ^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/uk_bin_collection/uk_bin_collection/collect_data.py", line 121, in client_code
    return get_bin_data_class.template_method(address_url, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/uk_bin_collection/uk_bin_collection/get_bin_data.py", line 61, in template_method
    bin_data_dict = self.get_and_parse_data(this_url, **kwargs)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/uk_bin_collection/uk_bin_collection/get_bin_data.py", line 82, in get_and_parse_data
    bin_data_dict = self.parse_data(page, url=address_url, **kwargs)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/uk_bin_collection/uk_bin_collection/councils/WestBerkshireCouncil.py", line 35, in parse_data
    driver = create_webdriver(web_driver, headless, None, __name__)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/uk_bin_collection/uk_bin_collection/common.py", line 326, in create_webdriver
    return webdriver.Chrome(
           ^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/selenium/webdriver/chrome/webdriver.py", line 45, in __init__
    super().__init__(
  File "/usr/local/lib/python3.12/site-packages/selenium/webdriver/chromium/webdriver.py", line 55, in __init__
    self.service.start()
  File "/usr/local/lib/python3.12/site-packages/selenium/webdriver/common/service.py", line 104, in start
    self._start_process(self._path)
  File "/usr/local/lib/python3.12/site-packages/selenium/webdriver/common/service.py", line 214, in _start_process
    self.process = subprocess.Popen(
                   ^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/subprocess.py", line 1026, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/local/lib/python3.12/subprocess.py", line 1955, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
OSError: [Errno 8] Exec format error: '/root/.wdm/drivers/chromedriver/linux64/114.0.5735.90/chromedriver'

```

### Verification

- [X] I searched for similar issues at https://github.com/robbrad/UKBinCollectionData/issues?q=is:issue and found no duplicates
- [X] If trying to add a specific council, I've checked it is listed as working at https://robbrad.github.io/UKBinCollectionData/3.12/
- [X] I have provided a detailed explanation of the issue as well as steps to replicate the issue
- [X] I understand that this project is run by volunteer contributors therefore completion of this issue cannot be guaranteed

**Comment by m26dvd** [comment]
@inspoadam When creating the service, was there a Selenium or local Chromium service available?

If it looked like this 
![image](https://github.com/user-attachments/assets/ef603669-9800-489e-ad96-1419ef36ccda)
then you need to either install Chromium or point to a remote Selenium server that is accessible

**Comment by inspoadam** [comment]
Thank @m26dvd. That is exactly what I am seeing. The problem the I am having now is that I have found zero help installing Chromium on my HA Yellow, so I am currently no further forward.

**Comment by inspoadam** [comment]
OK. So I now have it running. I had to jump through a fair number of hoops to get this working as I don't believe that there is a HACS package for Chromium/Selenium, so it was not a simple click and load approach if you have found your way here and still have no clue how to progress, then hopefully the following instructions will help.

FYI, I am running Home Assistant Yellow and while I could have installed Chromium/Selenium on a different machine, my preference was to run it on my Home Assistant box. I am aware that I may end up with resource issues and ultimately have to move it off, but for now it looks to be running OK and I am not expecting it to be heavily used. I also have plenty of RAM, so I am not expecting to run into issues.

The overview is that in order to get this working, I installed a docker image for Selenium, which was not straight forward, as Home Assistant Yellow uses a Raspberry Pi CM4, which is amd64 architecture. After a lot of Googling (most of it completely in the dark), I found the AMD image which made me realise what I needed to do and ultimately led me to [this](https://www.selenium.dev/blog/2024/multi-arch-images-via-docker-selenium/) link and I was finally starting to make some progress.

The first thing I needed to do was to be able to run a Docker image. I did this by using the **Advanced SSH & Web Terminal** Add On with Protected Mode turned off. If you have not used this before, there is a slight gotcha when running it for the first time which is solved byt adding a username and password in the options. I have added this as the error is not particularly intuitive and I was searching for a bit until I found the answer. Once you have this running, turn off Protection mode (located on the info tab) and restart the add-on.

<img width="553" alt="image" src="https://github.com/user-attachments/assets/75731bf5-e9b9-47ab-9719-d3882c735ded">

_Note: Before turning off protection mode, I strongly recommend taking a backup of your system as this is a very powerful mode and you can easily destroy your setup. I also recommend turning this off when you are done, just to be safe. If you do mess your system up, please do not blame me. :)_

Next head to the terminal where you can install Selenium. If you want a basic install, you can use the docker command provided in the above link, which is:

`docker run --rm -it -p 4444:4444 -p 5900:5900 -p 7900:7900 --shm-size 2g selenium/standalone-chromium:latest`

While will work, it is not perfect and it will firstly run in the foreground, secondly it will not automatically restart and thirdly, you will see a bunch of tracing errors in the logs (as there is no tracing service installed).

So, while progress has been made, it's not perfect. Long story short, I made some changes as the result is the following command that will create a container and will address all of the above issues. The command to do this is as follows:

`docker container create  --name selenium_standalone-chromium --restart always -it -e SE_ENABLE_TRACING=false -p 4444:4444 -p 5900:5900 -p 7900:7900 --shm-size 2g selenium/standalone-chromium:latest`

Start the container with the following:

`container start selenium_standalone-chromium`

Obviously, you can use your own naming convention if you don't like mine.

Assuming everything starts OK, then Selenium will now be installed and running. You can check this by running:

`docker stats`

Where you should see your containing running and consuming little resources after setting down from the initial startup. Just to be sure, check there are no errors in the logs by typing:

`docker logs selenium_standalone-chromium | more`

Obviously substituting your container name if using something different.

Assuming that everything is OK, you can now go back to the the **UK Bin Collection Data** integration and configure the service. If you got here after seeing the above screenshot, you should now see the following when setting it up.

<img width="298" alt="image" src="https://github.com/user-attachments/assets/a75d18e6-bef4-4920-8197-083f82069334">

All you have to do now is add the Remote Selenium Server URL as follows:

<img width="465" alt="image" src="https://github.com/user-attachments/assets/123f9c9c-9861-44c6-87f3-90733d749686">

...along with your own data (leaving the two Selenium server checkboxes as default) and you're done.

Finally, don't forget to **_turn off_** Protection mode for  Advanced SSH & Web Terminal.

**Comment by inspoadam** [comment]
@robbrad that would be my preferred option, but I have not been able to find a way to achieve this yet. I will keep trying as there is a fly in the ointment that my HA Supervisor is now complaining about and unsupported container on my system. 

If I have to stick with my current approach, my preference would be to install it outside of HAAS OS and directly on the HAY, but I have not yet worked out a way to get to this. If/when I make progress, I will post an update.

**Comment by inspoadam** [comment]
I have done some more digging and it turns out that when I do anything with Docker I am always working at the Hass OS level, so when I SSH directly into Hass OS, I see exactly the same Docker containers.

This means that I created the container at the highest level possible and also means that Supervisor is always going to complain about my "alien" container. It's not a big problem though as I just clicked on ignore and the error is no longer showing in my logs.

One option I do have is trying to turn this into a project published on HACS, but that is going to take a lots of reading, experimenting and head scratching by me.

In the meantime, if anyone knows how to install Selenium directly on a HAY, I am all ears.

**Comment by robbrad** [comment]
@inspoadam Selenium has two modes of execution- local driver or remote grid.

It's not that you need selenium- the integration has a selenium web driver in this integrations code.

This driver in the code, looks for Chrome or Firefox locally as a binary or if you set an IP it looks for a selenium grid, where an api command and runs on its chrome or Firefox binary.

Both use selenium from the integration, all you provide is a remote grid or local binary of chrome etc

With that said, if you can install chrome (chromium) or Firefox in the same place hass is running it will use it, when you select local. Just wanted that to be clear, you don't run selenium- you run a grid(docker) or the chrome binary

**Comment by robbrad** [comment]
Did you see this ? https://askubuntu.com/questions/1383789/install-chrome-on-ubuntu-debian-with-arm64

**Comment by inspoadam** [comment]
Thanks @robbrad. I did indeed find that and it's a great example of the rabbit hole I found myself in. When I found this, I thought that it would solve my problem, but unfortunately, I was unable to install is as HASS OS is not shipped with APT and it doesn't appear to be a straight forward install. At the same time, I really want to keep everything as standard as possible, and installing things outside of HASS is likely to cause me problems further down the line.

In the end, I felt that the docker approach was by far the cleanest, single box approach and it appear that the supervisor is only complaining about it as much as other HACS installs.

One thing that does come to mind. You mentioned above that the Selenium web driver is in the integration code. I have not been able to find that on my system and I am wondering if this was skipped due to failed compatibility checks. I am also wondering if finding the correct version of the driver and adding it in the right location will work?

ATEOTD, I may just install Chromium on a spare Raspberry Pi that I have hanging around and offbaord the entire process.
